3단계 - 질문 삭제하기 리팩터링

minseokLim opened the pull request:
1. PR 피드백에서 언급해주신 @Where을 적용하였습니다. 근데 편리한 만큼 부작용도 있는 것 같습니다. 예를 들면 질문을 삭제할 경우 거기에 귀속된 답변들도 모두 삭제 처리가 되어야 하다보니 Question 객체에 속한 answers 필드 내의 모든 Answer 객체의 deleted 값도 true로 변경됩니다. @Where(clause = "deleted = false") 라고 되어있다보니 개발자 입장에서는 '삭제되지 않은 값만 들고 있겠네?'라고 생각할 수 있고 거기에 맞춰 개발할 수도 있을 것 같습니다. 하지만 방금 예시와 같이 질문을 삭제할 경우, answers 필드 내에 삭제된 Answer 객체들이 들어있는 상황이 발생하게 됩니다. 사실 이 뿐만 아니라 이러한 상황은 다양하게 만들 수 있을 것 같습니다. 삭제된 Answer 객체를 Question 객체에 매핑해줄 경우에도 연관관계 편의 메서드에 의해 Question.answers 필드에 삭제된 Answer 객체가 포함될 수 있게될 것 같고요. 여기서 질문입니다.
  (1) 삭제나 Question 객체 매핑 등의 상황에서 Question.answers 필드에 삭제된 값이 들어가지 않도록 하기 위한 처리를 매번 해주는 것이 옳은 방법일까요, 아니면 애초에 삭제된 값만 들어있을 거라고 예상하는 개발자가 잘못된 생각을 하는 걸까요?
  (2) 삭제된 Answer도 조회를 해야하는 상황이 생기면 어떻게 수정을 하는게 좋을까요? 이는 1번 질문의 답변이 뭔지에 따라 달라질 것 같은데 만약 answers필드에 삭제된 값이 들어가지 않도록 매번 처리를 해주는 게 올바른 방법이라면, 저 같으면 answers -> visualAnswers 로 필드명을 변경하고 totalAnswers와 같이 별도의 필드를 추가할 것 같습니다.
  (3) 이럴바에 엔티티를 불변객체로 구성하는 게 나은 건가라는 생각도 듭니다. 근데 그러기엔 엔티티 객체 내부에 로직을 옮길 수가 없네요. 애초에 엔티티 클래스 내부에 로직이 들어가는 게 잘못인 걸까요? 엔티티는 불변으로 구성하고, 별도의 도메인 서비스 레이어를 만드는 편이 나으려나요?
 
2. QuestionTest.deleteQuestionHavingAnswersFromOthers()에서 이해할 수 없는 상황이 자꾸 발생합니다. 상황은 javajigi가 만든 Question에 javajigi의 Answer와 minseoklim의 Answer 2가지가 있는 상황인데요. javajigi의 Answer가 먼저 검사가 되기 때문에, 해당 객체의 delete = true로 변경됩니다. 트랜잭션 내부에서 익셉션이 발생하기 때문에 당연히 변경감지가 일어나지 않을 거라고 생각했는데 업데이트 쿼리가 날아가더라고요...-_-;;; 뭐가 문제인 걸까요?; 생각보다 JPA가 그만큼 똑똑하진 않은 걸까요? 만약 그렇다면 Answers객체 내부에서 try/catch 구문을 통해 원복 처리를 해줘야하는 걸까요? 그걸 기대하고 CannotDeleteException이 Checked Exception으로 되어있던 걸까요? 그것도 아니라면, Answers 내부에서 모든 권한 검사를 마친 후에 일괄적으로 객체의 상태를 변경해주는 게 맞는 걸까요? 
 여러가지 고민을 하다가 일단은 가장 마지막 방법, Answers 객체에서 모든 권한을 검사한 후 일괄적으로 객체 상태를 업데이트하는 방식으로 구현하였습니다. 근데 아무리 생각해도 익셉션이 발생했는데 더티 체킹이 일어난다는게 이상하긴 합니다.
 
3. 엔티티 내의 setter를 모두 제거하였습니다. 그 과정에서 학습테스트 메서드도 일부 삭제되었네요. 학습테스트 때문에 위험할 수도 있는 setter를 남겨놓는 건 좀 아닌 것 같다는 생각이 들었습니다. setter보단 기능이 추가될 때마다 의미있는 메시지를 던지는 것이 바람직하다고 생각합니다. 뿐만 아니라 당장 사용되지 않는 메서드나 클래스 등도 삭제하였습니다.

4. QuestionAnswers라는 일급컬렉션은 Question 객체에 종속될텐데 차라리 inner class로 넣는게 바람직할까요?

잘 부탁드립니다! :smile:

minseokLim committed:
refactor: PR 피드백 반영 (필드에 @Where 어노테이션을 추가하여 논리적으로 삭제되지 않은 엔티티들만 리스트에 포함되도록 수정)

minseokLim committed:
docs(README.md): 기능 목록 작성

minseokLim committed:
refactor: QnaService.deleteQuestion() 메서드 내의 비즈니스 로직을 도메인 객체 내부로 이동

QnaService.deleteQuestion() 메서드 내의 비즈니스 로직을 도메인 객체 내부로 이동
사용되지 않는 메서드/클래스 등 제거
엔티티 내 settter 제거

minseokLim committed:
feat: 엔티티 생성자에 not null 필드들에 대한 유효성 검사 추가

minseokLim committed:
refactor(User): 아이디와 비밀번호를 Account라는 임베디드 객체로 포장

minseokLim committed:
refactor(Answer, Question): contents와 User 객체를 Contents라는 임베디드 객체로 포장

minseokLim committed:
style: 린트 조건 일부 변경 및 린팅

nokchax commented on src/main/java/qna/domain/Account.java:
```diff
@@ -0,0 +1,47 @@
+import javax.persistence.Embeddable;
+
+@Embeddable
+public class Account {
```
별도로 추출하셨네요 👍

nokchax commented on src/main/java/qna/domain/Account.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    Account(final String userId, final String password) {
+        this.userId = Objects.requireNonNull(userId);
+        this.password = Objects.requireNonNull(password);
```
해당 메서드를 사용하게 되면 null인 객체가 파라미터로 넘어오는 경우 NPE가 발생하는데요,
NPE의 경우에는 스프링에서 500 에러를 리턴하게 되어있습니다. 500번대 에러는 서버측의 에러라고 볼 수 있는데, 이 로직의 경우 서버의 에러라기 보다는 요청이 잘못되었다는 걸 표현 해줘야 할 것같아요
NPE 사용은 경계해주세요!

minseokLim replied:
일반적인 웹서비스에서는 Request객체가 별도로 있고, 여기에서 유효성검사를 마치고 들어오기 때문에(여기서 유효성검사 실패시 400에러 발생) Entity까지 null로 들어왔다는 건 문제 상황으로 보는게 맞지 않을까요? 일반적으로 모니터링 시스템이 400번대는 잡지 않기 때문에 개발자가 상황 인지를 늦게할 수도 있을 것 같다는 생각이 드네요. 사실 이 부분에서는 굳이 Objects.requireNonNull를 추가하지 않더라도 디비컬럼이 not null이기 때문에 500이 발생하겠지만, 디비를 타기 이전에 '빠른 실패'라는 개념으로 추가해보았습니다. 의견 부탁드립니다😅

nokchax replied:
의도 하셨다면, 달리 이견은 없습니다.
그래도 NPE 사용은 지양하는게 좋아보여요 :)

nokchax commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,73 +1,73 @@
+
+        this.deleted = true;
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, id, loginUser));
+        return deleteHistories;
```
깔끔하게 리팩토링 진행해 주셨습니다 👍

nokchax commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,73 +1,73 @@
+        final List<DeleteHistory> deleteHistories = answers.deleteBy(loginUser);
+
+        this.deleted = true;
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, id, loginUser));
```
question의 작성자를 체크하는 로직, deleted 를 true로 셋하는 로직, deletehistory를 생성하는 로직은 하나의 메서드로 묶는 게 좋아보입니다.

nokchax commented on src/main/java/qna/domain/Question.java:
```diff
@@ -1,73 +1,73 @@
+        User writer = contents.getWriter();
+        if (!writer.equalsAccount(loginUser)) {
+            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        }
```
contents 로부터 writer를 꺼내오지 말고 loginUser를 전달해서 판단하는 건 어떨까요?

nokchax commented on src/main/java/qna/domain/QuestionAnswers.java:
```diff
@@ -0,0 +1,74 @@
+    private void checkAuthorities(final User loginUser) {
+        for (Answer answer : answers) {
+            answer.checkAuthority(loginUser);
+        }
```
authority 체크는 answer 클래스의 deleteBy 내부로 가는게 좋을거 같아요

nokchax commented on src/main/java/qna/domain/QuestionAnswers.java:
```diff
@@ -0,0 +1,74 @@
+
+    void add(final Answer answer) {
+        answers.add(answer);
+        validate();
```
보통은 validation 먼저 하고 값을 처리하지 않을까요?

minseokLim replied:
일반적으로는 앞쪽에 하지만, 인스턴스필드에 변경사항이 생기는 경우에는 변경을 가한 후 최종 값에 대한 유효성 검사를 진행하는 편입니다. 그게 더 확실해보여서요^^; '빠른 실패'라는 측면에서 앞쪽에 하는 편이 더 나을까요? 의견 부탁드립니다. 일단 수정은 하겠습니다! 😅

nokchax replied:
빠른 실패라는 관점 보다는, 할당 후에 검증을 했을때 할당에 큰 의미가 있을까 싶은 생각에 의견 드렸습니다 :)

minseokLim replied:
저는 인스턴스 필드에 대해서 습관적으로 유효성 검사를 뒤에 하는 버릇이 있는데, 생각해보니 그 위치가 생성자가 아니라 이런 메서드라면 위험할 수 있겠네요; 변경감지로 인해 업데이트가 될지도 모를 일이군요^^;

nokchax commented on src/main/java/qna/domain/QuestionAnswers.java:
```diff
@@ -0,0 +1,74 @@
+        validate();
+    }
+
+    private void validate() {
+        if (!answers.isEmpty()) {
+            final Question question = answers.get(0).getQuestion();
+            validateSameQuestion(question);
+        }
+    }
+
+    private void validateSameQuestion(final Question question) {
+        final boolean isAllSameQuestion = answers.stream()
+            .allMatch(answer -> answer.getQuestion().equals(question));
+        if (!isAllSameQuestion) {
+            throw new IllegalArgumentException("QuestionAnswers객체는 하나의 Question객체에 종속된 객체입니다.");
+        }
+    }
```
question과 answer가 양방향 관계가 되면서 로직이 추가 되었네요, 양방향 관계는 단방향 관계에 비해 신경쓸 것이 많아집니다.

minseokLim replied:
그럼 이 경우에는 일대다 단방향이 더 바람직할까요? 저도 그 생각을 하긴 했는데 추가적인 update문으로 발생하는 성능 이슈나 명시성 등의 관점에서 봤을 때 양방향이 더 바람직하다고 판단했습니다.
근데 제가 지금 테스트를 좀 해봤는데, 일대다 단방향의 경우에도 다에 해당하는 값(여기선 Answer)이 한 개씩만 추가될 때는 그래봐야 업데이트문 한 개가 더 날라가는 거네요? 
이 경우에 양방향과 일대다 단방향 중 어느 것이 더 바람직하다고 생각하시는지 의견 부탁드립니다. 일단 저는 일대다 단방향으로 수정해보도록 하겠습니다.

nokchax replied:
남겨주신 것 처럼 단방향으로 구성했을때와 양방향으로 구성했을때 각각의 관계가 가지는 장단점이 있죠.
물론 양방향으로 구성하는게 일반적이지만 이번 과제에서는 단방향으로도 큰 문제가 없기 때문에 
단방향과 양방향에 대해 좀 더 생각해보자는 차원에서 남겼던 거였습니다.
피드백을 확실히 전달드려야 했는데, 애매하게 남겼네요 죄송합니다. 🙇‍♂️

minseokLim replied:
아뇨. 괜찮습니다! 덕분에 양방향 단방향 모두 학습할 수 있었습니다!^^;

nokchax commented on src/main/java/qna/domain/QuestionAnswers.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private void validateSameQuestion(final Question question) {
+        final boolean isAllSameQuestion = answers.stream()
+            .allMatch(answer -> answer.getQuestion().equals(question));
+        if (!isAllSameQuestion) {
+            throw new IllegalArgumentException("QuestionAnswers객체는 하나의 Question객체에 종속된 객체입니다.");
+        }
```
```suggestion
        answers.stream()
                .filter(answer -> answer.getQuestion().equals(question))
                .findAny()
                .orElseThrow(() -> new IllegalArgumentException("QuestionAnswers객체는 하나의 Question객체에 종속된 객체입니다."));
```
스트림을 사용한 김에 이런식으로 해보는건 어떨까요

nokchax reviewed (changes requested):
안녕하세요, 민석님!
3단계 진행하느라 고생하셨습니다, 고민하신게 많이 느껴지네요 👍
피드백과 함께 질문 주신것에 대한 답변 남겨뒀어요, 확인후에 재요청 부탁드립니다 🙇‍♂️

nokchax commented:
고민을 많이 하셨네요 보기 좋습니다 👍 :)
1 - 1) 음 우선
```
answers 필드 내에 삭제된 Answer 객체들이 들어있는 상황이 발생하게 됩니다
```
라고 하셨는데 이건 기존의 `@Where` 없이도 동일하게 발생하는 것 아닌가요? `@where`을 사용해서 발생하는 문제라고 보기엔 어려움이 있는거 같아요. 보통 어떤 데이터를 삭제한다고 하면 삭제한 직후에는 해당 객체를 이용하지 않을 것 같습니다.
삭제나 객체 매핑에서 answers 필드에 삭제된 값이 들어가지 않도록 한다기 보다는 일단 삭제나 매핑을 진행하고 그 이후에 해당 객체를 다시 불러오는 식이 깔끔해 보입니다.

이번 과제를 진행하면서 어떠셨나요? `@Where` 없이 delete 내부의 로직을 수행하는데 더 많은 공수가 들어갈것 같지 않나요?
물론 `@Where`의 장단이 있겠지만 현재 상황에서는 적절하다고 판단해서 추천드린 것이구요, 무조건 적으로 사용하실 필요는 없습니다 :)

1 - 2) 
때에 따라 다를것 같은데요 말씀 주신 방법도 좋고, repository layer에서 메소드를 작성하고 answers만 가져오는 방식으로 하셔도 좋습니다. 이번 과정들은 객체지향을 연습하기 위한 것이니 만큼 너무 리뷰에 얽매이실 필요는 없어요, 때에 따라서는 repository layer를 사용하는게 좋은 방법이 되기도 합니다.

1 - 3)
불변객체는 내부의 모든 값이 변하지 않아야하는데, 우리가 서비스를 다루면서 그러한 경우가 얼마나 될까요? 로직은 도메인 내부에 두는 것이 맞습니다. 서비스 레이어에서(도메인 외부에서) 처리를 하면 getter사용이 불가피 하고 이는 객체지향이 아닌 절자지향이라 봐야겠지요.


2. 해당 테스트를 돌려봣는데 update쿼리는 따로 날아가지 않더라구요, 다른 테스트랑 헷갈리신 걸까요?
```
2021-11-16 20:27:44.148  INFO 44237 --- [           main] o.s.t.c.transaction.TransactionContext   : Began transaction (1) for test context [DefaultTestContext@247d8ae testClass = QuestionTest, testInstance = qna.domain.QuestionTest@520a95ff, testMethod = deleteQuestionHavingAnswersFromOthers@QuestionTest, testException = [null], mergedContextConfiguration = [MergedContextConfiguration@48974e45 testClass = QuestionTest, locations = '{}', classes = '{class qna.Application}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.OverrideAutoConfigurationContextCustomizerFactory$DisableAutoConfigurationContextCustomizer@478190fc, org.springframework.boot.test.autoconfigure.actuate.metrics.MetricsExportContextCustomizerFactory$DisableMetricExportContextCustomizer@1189dd52, org.springframework.boot.test.autoconfigure.filter.TypeExcludeFiltersContextCustomizer@351584c0, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@a6ecd19c, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@7d8995e, [ImportsContextCustomizer@6a84a97d key = [org.springframework.boot.autoconfigure.cache.CacheAutoConfiguration, org.springframework.boot.autoconfigure.data.jpa.JpaRepositoriesAutoConfiguration, org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceTransactionManagerAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.JdbcTemplateAutoConfiguration, org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration, org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration, org.springframework.boot.autoconfigure.transaction.TransactionAutoConfiguration, org.springframework.boot.test.autoconfigure.jdbc.TestDatabaseAutoConfiguration, org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManagerAutoConfiguration]], org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@639c2c1d, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@765d7657, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.context.SpringBootTestArgs@1, org.springframework.boot.test.context.SpringBootTestWebEnvironment@0], contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map['org.springframework.test.context.event.ApplicationEventsTestExecutionListener.recordApplicationEvents' -> false]]; transaction manager [org.springframework.orm.jpa.JpaTransactionManager@17c0274c]; rollback [true]
Hibernate: 
    insert 
    into
        user
        (id, created_at, updated_at, password, user_id, email, name) 
    values
        (null, ?, ?, ?, ?, ?, ?)
2021-11-16 20:27:44.150 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.149]
2021-11-16 20:27:44.150 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.149]
2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [password]
2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [VARCHAR] - [javajigi]
2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [VARCHAR] - [javajigi@slipp.net]
2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [VARCHAR] - [name]
Hibernate: 
    insert 
    into
        question
        (id, created_at, updated_at, contents, writer_id, deleted, title) 
    values
        (null, ?, ?, ?, ?, ?, ?)
2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.152]
2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.152]
2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [contents1]
2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [BIGINT] - [2]
2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [BOOLEAN] - [false]
2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [VARCHAR] - [title1]
Hibernate: 
    insert 
    into
        answer
        (id, created_at, updated_at, contents, writer_id, deleted, question_id) 
    values
        (null, ?, ?, ?, ?, ?, ?)
2021-11-16 20:27:44.179 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.176]
2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.176]
2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [이제는 잘 시간입니다]
2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [BIGINT] - [2]
2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [BOOLEAN] - [false]
2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [BIGINT] - [2]
Hibernate: 
    insert 
    into
        user
        (id, created_at, updated_at, password, user_id, email, name) 
    values
        (null, ?, ?, ?, ?, ?, ?)
2021-11-16 20:27:44.182 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.181]
2021-11-16 20:27:44.182 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.181]
2021-11-16 20:27:44.182 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [1234]
2021-11-16 20:27:44.183 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [VARCHAR] - [minseoklim]
2021-11-16 20:27:44.183 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [VARCHAR] - [mslim@slipp.net]
2021-11-16 20:27:44.183 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [VARCHAR] - [임민석]
Hibernate: 
    insert 
    into
        answer
        (id, created_at, updated_at, contents, writer_id, deleted, question_id) 
    values
        (null, ?, ?, ?, ?, ?, ?)
2021-11-16 20:27:44.184 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.184]
2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.184]
2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [이제는 잘 시간입니다]
2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [BIGINT] - [3]
2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [BOOLEAN] - [false]
2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [BIGINT] - [2]
Hibernate: 
    select
        answer0_.id as id1_0_,
        answer0_.created_at as created_2_0_,
        answer0_.updated_at as updated_3_0_,
        answer0_.contents as contents4_0_,
        answer0_.writer_id as writer_i6_0_,
        answer0_.deleted as deleted5_0_,
        answer0_.question_id as question7_0_ 
    from
        answer answer0_
Hibernate: 
    select
        question0_.id as id1_2_,
        question0_.created_at as created_2_2_,
        question0_.updated_at as updated_3_2_,
        question0_.contents as contents4_2_,
        question0_.writer_id as writer_i7_2_,
        question0_.deleted as deleted5_2_,
        question0_.title as title6_2_ 
    from
        question question0_
Hibernate: 
    select
        user0_.id as id1_3_,
        user0_.created_at as created_2_3_,
        user0_.updated_at as updated_3_3_,
        user0_.password as password4_3_,
        user0_.user_id as user_id5_3_,
        user0_.email as email6_3_,
        user0_.name as name7_3_ 
    from
        user user0_
2021-11-16 20:27:44.203  INFO 44237 --- [           main] o.s.t.c.transaction.TransactionContext   : Rolled back transaction for test: [DefaultTestContext@247d8ae testClass = QuestionTest, testInstance = qna.domain.QuestionTest@520a95ff, testMethod = deleteQuestionHavingAnswersFromOthers@QuestionTest, testException = [null], mergedContextConfiguration = [MergedContextConfiguration@48974e45 testClass = QuestionTest, locations = '{}', classes = '{class qna.Application}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.OverrideAutoConfigurationContextCustomizerFactory$DisableAutoConfigurationContextCustomizer@478190fc, org.springframework.boot.test.autoconfigure.actuate.metrics.MetricsExportContextCustomizerFactory$DisableMetricExportContextCustomizer@1189dd52, org.springframework.boot.test.autoconfigure.filter.TypeExcludeFiltersContextCustomizer@351584c0, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@a6ecd19c, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@7d8995e, [ImportsContextCustomizer@6a84a97d key = [org.springframework.boot.autoconfigure.cache.CacheAutoConfiguration, org.springframework.boot.autoconfigure.data.jpa.JpaRepositoriesAutoConfiguration, org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceTransactionManagerAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.JdbcTemplateAutoConfiguration, org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration, org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration, org.springframework.boot.autoconfigure.transaction.TransactionAutoConfiguration, org.springframework.boot.test.autoconfigure.jdbc.TestDatabaseAutoConfiguration, org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManagerAutoConfiguration]], org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@639c2c1d, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@765d7657, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.context.SpringBootTestArgs@1, org.springframework.boot.test.context.SpringBootTestWebEnvironment@0], contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map['org.springframework.test.context.event.ApplicationEventsTestExecutionListener.recordApplicationEvents' -> false]]

```

4. 
이건 선호도 차이일 것 같은데, 저는 클래스를 따로 빼두는 편입니다. 이너 클래스로 두면 인덴트도 늘어나고 클래스 내용이 길어져서 선호하지 않아요 :)

minseokLim committed:
docs(README.md): 기능 목록 구현 완료 체크박스 체크

minseokLim committed:
refactor: PR 피드백 반영 - Contents.isWrittenBy() 메서드 추가

Contents.isWrittenBy() 메서드 추가를 통해 해당 자원에 대한 권한을 판단하는 로직을 Contents 객체 내로 이동
기타 코드 리팩터링

minseokLim committed:
refactor: PR 피드백 반영 - Entity 내의 Objects.requireNonNull() 유효성 검사 제거

minseokLim commented:
> 고민을 많이 하셨네요 보기 좋습니다 👍 :) 1 - 1) 음 우선
> 
> ```
> answers 필드 내에 삭제된 Answer 객체들이 들어있는 상황이 발생하게 됩니다
> ```
> 
> 라고 하셨는데 이건 기존의 `@Where` 없이도 동일하게 발생하는 것 아닌가요? `@where`을 사용해서 발생하는 문제라고 보기엔 어려움이 있는거 같아요. 보통 어떤 데이터를 삭제한다고 하면 삭제한 직후에는 해당 객체를 이용하지 않을 것 같습니다. 삭제나 객체 매핑에서 answers 필드에 삭제된 값이 들어가지 않도록 한다기 보다는 일단 삭제나 매핑을 진행하고 그 이후에 해당 객체를 다시 불러오는 식이 깔끔해 보입니다.
> 
> 이번 과제를 진행하면서 어떠셨나요? `@Where` 없이 delete 내부의 로직을 수행하는데 더 많은 공수가 들어갈것 같지 않나요? 물론 `@Where`의 장단이 있겠지만 현재 상황에서는 적절하다고 판단해서 추천드린 것이구요, 무조건 적으로 사용하실 필요는 없습니다 :)
> 
> 1 - 2) 때에 따라 다를것 같은데요 말씀 주신 방법도 좋고, repository layer에서 메소드를 작성하고 answers만 가져오는 방식으로 하셔도 좋습니다. 이번 과정들은 객체지향을 연습하기 위한 것이니 만큼 너무 리뷰에 얽매이실 필요는 없어요, 때에 따라서는 repository layer를 사용하는게 좋은 방법이 되기도 합니다.
> 
> 1 - 3) 불변객체는 내부의 모든 값이 변하지 않아야하는데, 우리가 서비스를 다루면서 그러한 경우가 얼마나 될까요? 로직은 도메인 내부에 두는 것이 맞습니다. 서비스 레이어에서(도메인 외부에서) 처리를 하면 getter사용이 불가피 하고 이는 객체지향이 아닌 절자지향이라 봐야겠지요.
> 
> 1. 해당 테스트를 돌려봣는데 update쿼리는 따로 날아가지 않더라구요, 다른 테스트랑 헷갈리신 걸까요?
> 
> ```
> 2021-11-16 20:27:44.148  INFO 44237 --- [           main] o.s.t.c.transaction.TransactionContext   : Began transaction (1) for test context [DefaultTestContext@247d8ae testClass = QuestionTest, testInstance = qna.domain.QuestionTest@520a95ff, testMethod = deleteQuestionHavingAnswersFromOthers@QuestionTest, testException = [null], mergedContextConfiguration = [MergedContextConfiguration@48974e45 testClass = QuestionTest, locations = '{}', classes = '{class qna.Application}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.OverrideAutoConfigurationContextCustomizerFactory$DisableAutoConfigurationContextCustomizer@478190fc, org.springframework.boot.test.autoconfigure.actuate.metrics.MetricsExportContextCustomizerFactory$DisableMetricExportContextCustomizer@1189dd52, org.springframework.boot.test.autoconfigure.filter.TypeExcludeFiltersContextCustomizer@351584c0, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@a6ecd19c, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@7d8995e, [ImportsContextCustomizer@6a84a97d key = [org.springframework.boot.autoconfigure.cache.CacheAutoConfiguration, org.springframework.boot.autoconfigure.data.jpa.JpaRepositoriesAutoConfiguration, org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceTransactionManagerAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.JdbcTemplateAutoConfiguration, org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration, org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration, org.springframework.boot.autoconfigure.transaction.TransactionAutoConfiguration, org.springframework.boot.test.autoconfigure.jdbc.TestDatabaseAutoConfiguration, org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManagerAutoConfiguration]], org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@639c2c1d, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@765d7657, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.context.SpringBootTestArgs@1, org.springframework.boot.test.context.SpringBootTestWebEnvironment@0], contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map['org.springframework.test.context.event.ApplicationEventsTestExecutionListener.recordApplicationEvents' -> false]]; transaction manager [org.springframework.orm.jpa.JpaTransactionManager@17c0274c]; rollback [true]
> Hibernate: 
>     insert 
>     into
>         user
>         (id, created_at, updated_at, password, user_id, email, name) 
>     values
>         (null, ?, ?, ?, ?, ?, ?)
> 2021-11-16 20:27:44.150 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.149]
> 2021-11-16 20:27:44.150 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.149]
> 2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [password]
> 2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [VARCHAR] - [javajigi]
> 2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [VARCHAR] - [javajigi@slipp.net]
> 2021-11-16 20:27:44.151 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [VARCHAR] - [name]
> Hibernate: 
>     insert 
>     into
>         question
>         (id, created_at, updated_at, contents, writer_id, deleted, title) 
>     values
>         (null, ?, ?, ?, ?, ?, ?)
> 2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.152]
> 2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.152]
> 2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [contents1]
> 2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [BIGINT] - [2]
> 2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [BOOLEAN] - [false]
> 2021-11-16 20:27:44.153 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [VARCHAR] - [title1]
> Hibernate: 
>     insert 
>     into
>         answer
>         (id, created_at, updated_at, contents, writer_id, deleted, question_id) 
>     values
>         (null, ?, ?, ?, ?, ?, ?)
> 2021-11-16 20:27:44.179 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.176]
> 2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.176]
> 2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [이제는 잘 시간입니다]
> 2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [BIGINT] - [2]
> 2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [BOOLEAN] - [false]
> 2021-11-16 20:27:44.180 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [BIGINT] - [2]
> Hibernate: 
>     insert 
>     into
>         user
>         (id, created_at, updated_at, password, user_id, email, name) 
>     values
>         (null, ?, ?, ?, ?, ?, ?)
> 2021-11-16 20:27:44.182 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.181]
> 2021-11-16 20:27:44.182 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.181]
> 2021-11-16 20:27:44.182 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [1234]
> 2021-11-16 20:27:44.183 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [VARCHAR] - [minseoklim]
> 2021-11-16 20:27:44.183 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [VARCHAR] - [mslim@slipp.net]
> 2021-11-16 20:27:44.183 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [VARCHAR] - [임민석]
> Hibernate: 
>     insert 
>     into
>         answer
>         (id, created_at, updated_at, contents, writer_id, deleted, question_id) 
>     values
>         (null, ?, ?, ?, ?, ?, ?)
> 2021-11-16 20:27:44.184 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [1] as [TIMESTAMP] - [2021-11-16T20:27:44.184]
> 2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [2] as [TIMESTAMP] - [2021-11-16T20:27:44.184]
> 2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [3] as [VARCHAR] - [이제는 잘 시간입니다]
> 2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [4] as [BIGINT] - [3]
> 2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [5] as [BOOLEAN] - [false]
> 2021-11-16 20:27:44.185 TRACE 44237 --- [           main] o.h.type.descriptor.sql.BasicBinder      : binding parameter [6] as [BIGINT] - [2]
> Hibernate: 
>     select
>         answer0_.id as id1_0_,
>         answer0_.created_at as created_2_0_,
>         answer0_.updated_at as updated_3_0_,
>         answer0_.contents as contents4_0_,
>         answer0_.writer_id as writer_i6_0_,
>         answer0_.deleted as deleted5_0_,
>         answer0_.question_id as question7_0_ 
>     from
>         answer answer0_
> Hibernate: 
>     select
>         question0_.id as id1_2_,
>         question0_.created_at as created_2_2_,
>         question0_.updated_at as updated_3_2_,
>         question0_.contents as contents4_2_,
>         question0_.writer_id as writer_i7_2_,
>         question0_.deleted as deleted5_2_,
>         question0_.title as title6_2_ 
>     from
>         question question0_
> Hibernate: 
>     select
>         user0_.id as id1_3_,
>         user0_.created_at as created_2_3_,
>         user0_.updated_at as updated_3_3_,
>         user0_.password as password4_3_,
>         user0_.user_id as user_id5_3_,
>         user0_.email as email6_3_,
>         user0_.name as name7_3_ 
>     from
>         user user0_
> 2021-11-16 20:27:44.203  INFO 44237 --- [           main] o.s.t.c.transaction.TransactionContext   : Rolled back transaction for test: [DefaultTestContext@247d8ae testClass = QuestionTest, testInstance = qna.domain.QuestionTest@520a95ff, testMethod = deleteQuestionHavingAnswersFromOthers@QuestionTest, testException = [null], mergedContextConfiguration = [MergedContextConfiguration@48974e45 testClass = QuestionTest, locations = '{}', classes = '{class qna.Application}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTestContextBootstrapper=true}', contextCustomizers = set[org.springframework.boot.test.autoconfigure.OverrideAutoConfigurationContextCustomizerFactory$DisableAutoConfigurationContextCustomizer@478190fc, org.springframework.boot.test.autoconfigure.actuate.metrics.MetricsExportContextCustomizerFactory$DisableMetricExportContextCustomizer@1189dd52, org.springframework.boot.test.autoconfigure.filter.TypeExcludeFiltersContextCustomizer@351584c0, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@a6ecd19c, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@7d8995e, [ImportsContextCustomizer@6a84a97d key = [org.springframework.boot.autoconfigure.cache.CacheAutoConfiguration, org.springframework.boot.autoconfigure.data.jpa.JpaRepositoriesAutoConfiguration, org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.DataSourceTransactionManagerAutoConfiguration, org.springframework.boot.autoconfigure.jdbc.JdbcTemplateAutoConfiguration, org.springframework.boot.autoconfigure.liquibase.LiquibaseAutoConfiguration, org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration, org.springframework.boot.autoconfigure.transaction.TransactionAutoConfiguration, org.springframework.boot.test.autoconfigure.jdbc.TestDatabaseAutoConfiguration, org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManagerAutoConfiguration]], org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@639c2c1d, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@765d7657, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.context.SpringBootTestArgs@1, org.springframework.boot.test.context.SpringBootTestWebEnvironment@0], contextLoader = 'org.springframework.boot.test.context.SpringBootContextLoader', parent = [null]], attributes = map['org.springframework.test.context.event.ApplicationEventsTestExecutionListener.recordApplicationEvents' -> false]]
> ```
> 
> 
> 이건 선호도 차이일 것 같은데, 저는 클래스를 따로 빼두는 편입니다. 이너 클래스로 두면 인덴트도 늘어나고 클래스 내용이 길어져서 선호하지 않아요 :)

1. 일단 @Where 사용하니까 너무 편했습니다ㅋㅋ 답변 감사합니다. 현재 회사에서 soft delete를 씀에도 불구하고 삭제되지 않은 것과 삭제여부 상관없이 모두 조회하는 상황이 많다보니, 요구사항이 바뀌었을 때 어떻게 대응해야할 지 궁금했습니다. 로직은 도메인 내부에 있어야한다는 말씀, 정말 공감됩니다. 현재 회사에서 getter를 통해 로직을 구현하면서 고통받고 있습니다...;;
2. 이건 제가 구현을 바꾸면서 결국 통과시키게 만들어서 그런겁니다^^; 위에 코멘트 달아주신 내용들 중 제가 Resolve처리하지 않은 내용들이 다 이 테스트 케이스와 관련이 있는데요. 분석하다보니 원인을 찾았습니다!! 코드에 주석달아놨으니 확인해주시고 다시 Change Request주시면 삭제 후 다시 PR 요청하도록 하겠습니다! 😅 
4. 넵! 그럼 저도 그렇게 하지 않도록 하겠습니다ㅋㅋ

minseokLim committed:
refactor: PR 피드백 반영 - deleteBy() 메서드 리팩터링

nokchax commented on src/main/java/qna/domain/Question.java:
```diff
@@ -42,13 +43,16 @@ public Question(final Long id, final User writer, final String title, final Stri
-        checkAuthority(loginUser);
-
-        final List<DeleteHistory> deleteHistories = answers.deleteBy(loginUser);
+        final List<DeleteHistory> deleteHistories = new ArrayList<>();
```
피드백 잘 반영해 주셨습니다 👍

nokchax commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -122,7 +122,37 @@ void deleteQuestionHavingAnswersFromOthers() {
+        */
+
+        // 아래 주석을 풀어보시면 update 쿼리를 확인하실 수 있습니다.
+        // assertThat(question.isDeleted()).isFalse();
+        // assertThat(answer.isDeleted()).isFalse();
```
테스트를 해봤는데요, 정확히는 같은 클래스 내에 존재하는 `@AfterEach`로 실행되는 tearDown() 때문인거 같아요
```
assertThat(question.isDeleted()).isFalse();
```
이때 객체의 값은 `true`이기 때문에 해당 검증은 실패하게 되고 테스트가 종료되겠죠.
종료된 이후에 tearDown() 내부의 deleteAll() 로직이 수행되기 전에 PersistenceContext에 변경된 작업들을 
클리어 하는 작업이 발생하면서 update쿼리가 발생하게 된 것 같습니다.
실제로 해당 주석을 풀고 tearDown()내의 deleteAll() 을 주석처리하면 update쿼리가 발생 안하는 것을 보실 수 있습니다 :)

여담이지만 assertThat으로 출력되는 로그가 update 쿼리보다 더 나중에 뜨는 것은 해당 로그가 `@AfterEach` 이후에 출력되도록 만들어져 있어서 그런것 같습니다.

nokchax replied:
코멘트로 남겨주신 질문에도 답을 남겼습니다 확인해주세요 :)

minseokLim replied:
친절하고 자세한 답변 감사합니다! 😄

nokchax reviewed (changes requested):
안녕하세요, 민석님! 많은 고민들을 하시는걸 보고 저도 많이 자극 받는거 같아요 🙇‍♂️ 
질문에 답 남겼습니다. 확인후에 재요청 주시면 바로 머지하도록 할게요 :)

minseokLim committed:
refactor: 테스트 코드 주석 제거

nokchax reviewed (approved):
JPA 과제 진행하느라 고생 많으셨습니다 👏👏👏 
민석님이 질문을 많이해 주셔서 저도 많은 공부를 했던거 같아요 :)
남은 과정도 파이팅하시고, 꼭 완주하기를 응원하겠습니다 💪

minseokLim commented:
> JPA 과제 진행하느라 고생 많으셨습니다 👏👏👏 민석님이 질문을 많이해 주셔서 저도 많은 공부를 했던거 같아요 :) 남은 과정도 파이팅하시고, 꼭 완주하기를 응원하겠습니다 💪

감사합니다! 꼭 완주하도록 하겠습니다! 물론 많이 배우면서요! 😄
